File: 2016/13/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	input, _ := strconv.Atoi(string(data))
	fmt.Println(part1(input))
	fmt.Println(part2(input))
}

func mapout(coord [2]int, layout *map[[2]int]bool, input int) bool {
	if _, ok := (*layout)[coord]; !ok {
		sum := coord[0]*coord[0] + 3*coord[0] + 2*coord[0]*coord[1] + coord[1] + coord[1]*coord[1] + input
		count := 0
		for sum > 0 {
			count += sum & 1
			sum >>= 1
		}
		if count%2 == 1 {
			(*layout)[coord] = false
		} else {
			(*layout)[coord] = true
		}
	}
	return (*layout)[coord]
}

func part1(input int) int {
	queue := [][3]int{{1, 1, 0}}
	steps := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	layout := map[[2]int]bool{}
	visited := map[[2]int]int{}
	layout[[2]int{1, 1}] = true
	visited[[2]int{1, 1}] = 0
	i := 0
	for i < len(queue) {
		current := queue[i]
		pos := [2]int{current[0], current[1]}
		dist := current[2]
		for _, step := range steps {
			coord := [2]int{pos[0] + step[0], pos[1] + step[1]}
			if coord[0] < 0 || coord[1] < 0 {
				continue
			}
			if coord[0] == 31 && coord[1] == 39 {
				return dist + 1
			}
			if mapout(coord, &layout, input) {
				d, ok := visited[coord]
				if !ok || dist+1 < d {
					queue = append(queue, [3]int{coord[0], coord[1], dist + 1})
					visited[coord] = dist + 1
				}
			}
		}
		i++
	}
	return -1
}

func part2(input int) int {
	queue := [][3]int{{1, 1, 0}}
	steps := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
	layout := map[[2]int]bool{}
	visited := map[[2]int]int{}
	layout[[2]int{1, 1}] = true
	visited[[2]int{1, 1}] = 0
	i := 0
	for i < len(queue) {
		current := queue[i]
		if current[2] < 50 {
			pos := [2]int{current[0], current[1]}
			dist := current[2]
			for _, step := range steps {
				coord := [2]int{pos[0] + step[0], pos[1] + step[1]}
				if coord[0] < 0 || coord[1] < 0 {
					continue
				}
				if mapout(coord, &layout, input) {
					d, ok := visited[coord]
					if !ok || dist+1 < d {
						queue = append(queue, [3]int{coord[0], coord[1], dist + 1})
						visited[coord] = dist + 1
					}
				}
			}
		}
		i++
	}
	return len(visited)
}

File: 2016/13/solution.ts
import * as fs from "fs";
import * as path from "path";
import MinHeap from "../../utils/MinHeap";

const puzzleInput = Number(
  fs.readFileSync(path.join(__dirname, "input.txt"), "utf-8")
);

const directions = [
  [0, 1],
  [1, 0],
  [0, -1],
  [-1, 0],
];

function isOpenSpace(x: number, y: number, puzzleInput: number): boolean {
  return (
    (x * x + 3 * x + 2 * x * y + y + y * y + puzzleInput)
      .toString(2)
      .replaceAll("0", "").length %
      2 ===
    0
  );
}

interface State {
  steps: number;
  pos: [number, number];
}

function part1(puzzleInput: number): number {
  const xGoal = 31;
  const yGoal = 39;

  const queue: [number, State][] = [[0, { steps: 0, pos: [1, 1] }]];
  const visited: Map<string, number> = new Map([["1,1", 0]]);

  function dist(x: number, y: number): number {
    return Math.abs(x - xGoal) + Math.abs(y - yGoal);
  }

  while (queue.length) {
    let [_, { steps, pos }] = MinHeap.pop(queue) as [number, State];
    const [x, y] = pos;
    if (x === xGoal && y === yGoal) return steps;
    steps++;

    for (const [dx, dy] of directions) {
      const [x2, y2] = [x + dx, y + dy];
      if (x2 < 0 || y2 < 0) continue;
      if (!isOpenSpace(x2, y2, puzzleInput)) continue;
      const coord = `${x2},${y2}`;
      if (visited.has(coord) && visited.get(coord)! <= steps) continue;
      visited.set(coord, steps);
      MinHeap.push(queue, [dist(x2, y2) + steps, { steps, pos: [x2, y2] }]);
    }
  }

  return -1;
}

function part2(puzzleInput: number): number {
  const queue: State[] = [{ steps: 0, pos: [1, 1] }];
  const visited: Map<string, number> = new Map([["1,1", 0]]);

  while (queue.length) {
    let { steps, pos } = queue.pop()!;
    if (steps === 50) continue;
    const [x, y] = pos;
    steps++;
    for (const [dx, dy] of directions) {
      const [x2, y2] = [x + dx, y + dy];
      if (x2 < 0 || y2 < 0) continue;
      if (!isOpenSpace(x2, y2, puzzleInput)) continue;
      const coord = `${x2},${y2}`;
      if (visited.has(coord) && visited.get(coord)! <= steps) continue;
      visited.set(coord, steps);
      queue.push({ steps, pos: [x2, y2] });
    }
  }

  return visited.size;
}

console.log(part1(puzzleInput));
console.log(part2(puzzleInput));
